File: sdk/php/src/Providers/MailGolem.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * MailGolem 渠道实现 — https://mailgolem.com
 *
 * Laravel session + CSRF：
 *   GET / 建立 session 并从隐藏 input 提取 CSRF；
 *   GET /random-email-address 获取随机邮箱；token 存 base64(JSON{csrf, cookies})。
 *   getEmails 重新拉首页刷新 session/CSRF，再 POST /fetch-emails/{email} 拉列表。
 */
final class MailGolem
{
    private const BASE = 'https://mailgolem.com';
    private const TOK_PREFIX = 'mgolem1:';

    private const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0';

    /**
     * @return array<string,string>
     */
    private static function defaultHeaders(): array
    {
        return [
            'Accept' => 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8',
            'Accept-Language' => 'en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7',
            'Cache-Control' => 'no-cache',
            'DNT' => '1',
            'Pragma' => 'no-cache',
            'User-Agent' => self::UA,
        ];
    }

    /**
     * 从响应 Set-Cookie 提取 cookie 映射（仅 name=value）。
     *
     * @return array<string,string>
     */
    private static function responseCookies(\Psr\Http\Message\ResponseInterface $resp): array
    {
        $out = [];
        foreach ($resp->getHeader('Set-Cookie') as $sc) {
            $first = trim(explode(';', $sc, 2)[0]);
            if ($first === '' || !str_contains($first, '=')) {
                continue;
            }
            [$k, $v] = explode('=', $first, 2);
            $k = trim($k);
            if ($k !== '') {
                $out[$k] = trim($v);
            }
        }
        return $out;
    }

    /**
     * @param array<string,string> $cookies
     */
    private static function cookieHeader(array $cookies): string
    {
        $parts = [];
        foreach ($cookies as $k => $v) {
            $parts[] = $k . '=' . $v;
        }
        return implode('; ', $parts);
    }

    private static function extractCsrf(string $html): string
    {
        if (preg_match('/<input[^>]+name="_token"[^>]+value="([^"]+)"/', $html, $m) !== 1) {
            throw new \RuntimeException('mailgolem: 无法从页面中提取 CSRF token');
        }
        return $m[1];
    }

    /**
     * @param array<string,string> $cookies
     */
    private static function encodeToken(string $csrf, array $cookies): string
    {
        $payload = json_encode(['csrf' => $csrf, 'c' => $cookies], JSON_UNESCAPED_SLASHES);
        return self::TOK_PREFIX . base64_encode((string) $payload);
    }

    public static function generate(): EmailInfo
    {
        $r1 = HttpClient::get(self::BASE . '/', self::defaultHeaders());
        $cookies = self::responseCookies($r1);
        $csrf = self::extractCsrf((string) $r1->getBody());

        $r2 = HttpClient::get(
            self::BASE . '/random-email-address',
            self::defaultHeaders() + ['Referer' => self::BASE . '/', 'Cookie' => self::cookieHeader($cookies)],
        );
        $cookies = array_merge($cookies, self::responseCookies($r2));
        $emailAddr = trim((string) $r2->getBody());
        if ($emailAddr === '' || !str_contains($emailAddr, '@')) {
            throw new \RuntimeException('mailgolem: 获取到无效的邮箱地址');
        }

        return new EmailInfo('mailgolem', $emailAddr, self::encodeToken($csrf, $cookies));
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email, ?string $token): array
    {
        if ($token === null || $token === '') {
            throw new \InvalidArgumentException('mailgolem: token 不能为空');
        }
        $addr = trim($email);
        if ($addr === '') {
            throw new \InvalidArgumentException('mailgolem: 邮箱地址不能为空');
        }

        // 原 session 可能已过期，重新获取 session 与 CSRF
        $r1 = HttpClient::get(self::BASE . '/', self::defaultHeaders());
        $cookies = self::responseCookies($r1);
        $newCsrf = self::extractCsrf((string) $r1->getBody());

        $r2 = HttpClient::post(
            self::BASE . '/fetch-emails/' . rawurlencode($addr),
            [
                'Accept' => 'application/json, text/plain, */*',
                'Accept-Language' => 'en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7',
                'Content-Type' => 'application/x-www-form-urlencoded',
                'X-Requested-With' => 'XMLHttpRequest',
                'Referer' => self::BASE . '/',
                'User-Agent' => self::UA,
                'Cookie' => self::cookieHeader($cookies),
            ],
            body: '_token=' . rawurlencode($newCsrf),
        );
        $body = HttpClient::json($r2);
        if (!array_is_list($body)) {
            return [];
        }

        $out = [];
        foreach ($body as $item) {
            if (!is_array($item)) {
                continue;
            }
            $raw = [
                'id' => (string) ($item['id'] ?? ''),
                'from' => $item['from'] ?? '',
                'to' => $addr,
                'subject' => $item['subject'] ?? '',
            ];
            $out[] = Normalize::email($raw, $addr);
        }
        return $out;
    }
}

File: sdk/csharp/src/Providers/Mailgolem.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace XxxXTeam.TempMail.Providers;

/// <summary>
/// MailGolem 渠道 — https://mailgolem.com
/// 基于 Laravel session + CSRF token。
/// 创建：GET / 取 session cookie + CSRF（input name="_token"）→ GET /random-email-address 取邮箱。
/// 收信：GET / 重取新 session + CSRF → POST /fetch-emails/{email}（body _token=）取列表。
/// token 编码 "mgolem1:" + base64(JSON{csrf,c=cookie})。
/// </summary>
public static class Mailgolem
{
    private const string Base = "https://mailgolem.com";
    private const string TokPrefix = "mgolem1:";

    private static readonly Regex CsrfRe = new("<input[^>]+name=\"_token\"[^>]+value=\"([^\"]+)\"", RegexOptions.Compiled);

    private static Dictionary<string, string> DefaultHeaders() => new()
    {
        ["Accept"] = "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
        ["Accept-Language"] = "en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7",
        ["Cache-Control"] = "no-cache",
        ["DNT"] = "1",
        ["Pragma"] = "no-cache",
        ["User-Agent"] = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
    };

    private static string ExtractCsrf(string html)
    {
        var m = CsrfRe.Match(html);
        if (!m.Success) throw new Exception("mailgolem: 无法从页面中提取 CSRF token");
        return m.Groups[1].Value;
    }

    private static string EncodeToken(string csrf, string cookie)
    {
        var raw = Json.Serialize(new Dictionary<string, object?> { ["csrf"] = csrf, ["c"] = cookie });
        return TokPrefix + Convert.ToBase64String(Encoding.UTF8.GetBytes(raw));
    }

    /// <summary>创建 mailgolem 临时邮箱</summary>
    public static EmailInfo Generate()
    {
        var r1 = Http.Get($"{Base}/", DefaultHeaders());
        r1.EnsureSuccess();
        var cookie = ProviderHttpUtil.CookieHeaderFrom(r1);
        var csrf = ExtractCsrf(r1.Body);

        var headers = DefaultHeaders();
        headers["Referer"] = $"{Base}/";
        if (!string.IsNullOrEmpty(cookie)) headers["Cookie"] = cookie;
        var r2 = Http.Get($"{Base}/random-email-address", headers);
        r2.EnsureSuccess();
        cookie = ProviderHttpUtil.MergeCookies(cookie, r2);

        var emailAddr = r2.Body.Trim();
        if (emailAddr.Length == 0 || !emailAddr.Contains('@'))
            throw new Exception($"mailgolem: 获取到无效的邮箱地址: {emailAddr}");

        return new EmailInfo("mailgolem", emailAddr, EncodeToken(csrf, cookie));
    }

    /// <summary>获取 mailgolem 邮件列表</summary>
    public static List<Email> GetEmails(string? token, string email)
    {
        if (string.IsNullOrEmpty(token)) throw new Exception("mailgolem: token 不能为空");
        var addr = (email ?? "").Trim();
        if (addr.Length == 0) throw new Exception("mailgolem: 邮箱地址不能为空");

        // 重新获取 session 和 CSRF（原 session 可能已过期）
        var r1 = Http.Get($"{Base}/", DefaultHeaders());
        r1.EnsureSuccess();
        var cookie = ProviderHttpUtil.CookieHeaderFrom(r1);
        var newCsrf = ExtractCsrf(r1.Body);

        var headers = new Dictionary<string, string>
        {
            ["Accept"] = "application/json, text/plain, */*",
            ["Accept-Language"] = "en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7",
            ["Content-Type"] = "application/x-www-form-urlencoded",
            ["X-Requested-With"] = "XMLHttpRequest",
            ["Referer"] = $"{Base}/",
            ["User-Agent"] = DefaultHeaders()["User-Agent"],
        };
        if (!string.IsNullOrEmpty(cookie)) headers["Cookie"] = cookie;

        var r2 = Http.Post($"{Base}/fetch-emails/{addr}", $"_token={newCsrf}",
            "application/x-www-form-urlencoded", headers);
        r2.EnsureSuccess();

        var result = new List<Email>();
        if (Json.Parse(r2.Body) is not JsonArray body) return result;

        foreach (var node in body)
        {
            if (node is not JsonObject item) continue;
            result.Add(Normalize.NormalizeEmail(new Dictionary<string, object?>
            {
                ["id"] = Json.Str(item, "id"),
                ["from"] = Json.Str(item, "from"),
                ["to"] = addr,
                ["subject"] = Json.Str(item, "subject"),
            }, addr));
        }
        return result;
    }
}
